File: practice/algorithms/strings/sherlock_and_valid_string/sherlock_and_valid_string.py
from collections import Counter

freqs = Counter(list(raw_input()))
s = set(freqs.values())

f = sorted(freqs.values())
if len(s) == 1:
    print 'YES'
elif len(s) > 2:
    print 'NO'
elif f[-1] - f[-2] == 1:
    print 'YES'
elif f[0] == 1 and f[1] != 1:
    print 'YES'
else:
    print 'NO'

File: practice/algorithms/strings/sherlock_and_valid_string/sherlock_and_valid_string.hs
import Control.Monad (forM_)
import qualified Data.Map as Map
import Data.Maybe
import Data.List
import qualified Data.Set as Set

addKey :: Char -> (Map.Map Char Int) -> (Map.Map Char Int)
addKey k m = Map.insertWith (+) k 1 m

counter :: String -> (Map.Map Char Int)
counter s = foldl (\acc x -> addKey x acc) Map.empty s

sortedFreq :: (Map.Map Char Int) -> [Int]
sortedFreq c = sort $ Map.foldr (\x acc -> x : acc) [] c

solve :: [Int] -> (Set.Set Int) -> String
solve f d
    | Set.size d == 1                    = "YES"
    | Set.size d  > 2                    = "NO"
    | f !! (n - 1) - (f !! (n - 2)) == 1 = "YES"
    | f !! 0 == 1 && f !! 1 > 1          = "YES"
    | otherwise                          = "NO" where
        n = length f

main = do
    s <- getLine
    let c = counter s
        f = sortedFreq c
        d = Set.fromList f
    putStrLn $ solve f d
